### DISCUSSION TITLE
Typesetting different kinds of text

### DISCUSSION THREAD

**Discussion by sasja-san** [quarto-dev/quarto-cli#10349] [Q&A]
### Description

I want to have different kinds of text in a document be typeset in different ways;
For a paragraph I want the first lines be black text, started by a bold, and then I want the next line to be red, started by a bold red word.

But maybe not red? I won't know until I've seen it in print. My guess is the code would look something like this

```qmd

|
| [Start of line is bold or something]{.start_a}[ and the rest normal]{.norm_a}.
| [Line nr two in paragraph is perhaps colored?]{.start_b}[ and the rest normal, but colored, probably]{.norm_b}.
|
```

I just borrowed the syntax from {.aside} here, but I hope you get my point. Any advice or pointers on how toying around with styles and testing could be done would be greatly appreciated. The end goal is a printed palmphlet, so maybe this task is easiest to do in latex directly?

**Comment by mcanouil** [comment]
Depending on the format, the components inside the curly brackets varies, *e.g.*, for HTML `style="color:red;"`.

See <https://m.canouil.dev/quarto-highlight-text/> for an example of implementation for multi format with a unified input syntax.

Regarding the "first sentence" etc. you need a LUA filter and define the logic to read/parse paragraphs, then apply what you want.

**Reply by sasja-san** [reply]
This looks nice. Thanks!

Is there some way to lift out the actual coloring and only have it declared only once? So instead of
```qmd
[Red text]{colour="#b22222"}
```
it would be nice with either of these
```qmd
[Red text]{colour=RED_COL_STRING}
[Red text]{RED_TEXT_STYLE}
```
(so I can easily do changes to all text when mucking about, trying to find *just* the right typesetting - is this what the LUA filter would do?)

**Reply by cderv** [reply]
You could have a Lua filter catching the attribute in the Span, and then change by your color definition. 

This should be possible more easily in the near feature through some YAML configuration to define colors. 

@cscheid @gordonwoodhull we discuss this recently, with some decision to make on the syntax. For inline, span attribute for brand color seems interesting ! 🤔

**Reply by gordonwoodhull** [reply]
There are starting to be branding features in 1.6, but I wouldn't count on them yet.

In the meantime, one approach is to use shortcodes:

````md
---
title: 'shortcode a color'
red: '#f11'
blue: '#11f'
green: '#1a1'
---

We want to color the words [red]{style='color: {{<meta red>}}'}, [blue]{style='color: {{<meta blue>}}'},
and [green]{style='color: {{<meta green>}}'}.
````

**Reply by mcanouil** [reply]
If I'm not mistaken, the shortcode needs to be inside single quote when inside attribute (as in the example). Double quotes won't work.

**Reply by cderv** [reply]
> In the meantime, one approach is to use shortcodes:

Nice ! Easier than my Lua filter idea indeed ! Thanks !

**Reply by mcanouil** [reply]
@gordonwoodhull does the style attribute also work for Typst with the CSS feature? (it won't for LaTex or OpenXML a.k.a. Word)

**Reply by gordonwoodhull** [reply]
> also work for Typst with the CSS feature?

Good point, my example only work for HTML and Typst.

I think you could use shortcodes in combination with @mcanouil's extension if you are targeting Latex or Docx.

**Reply by gordonwoodhull** [reply]
Using `highlight-text` + shortcodes:

````md
---
title: 'shortcode a color'
filters:
  - quarto
  - highlight-text
red: '#ff1111'
blue: '#1111ff'
green: '#11aa11'
---

We want to color the words [red]{color='{{<meta red>}}'}, [blue]{color='{{<meta blue>}}'}, and [green]{color='{{<meta green>}}'}.
````

Works across HTML, PDF, Typst and Docx.

**Reply by mcanouil** [reply]
Maybe it has to do with filter orders here. `highlight-text` should happen after Quarto because I did not write the Lua to handle shortcodes.

Additionally, you should use full hex colour code for now, that's the only colour code fully supported for all formats.

**Reply by gordonwoodhull** [reply]
Thanks, I've edited my answer!
